src/headers/frame_header.rs

```diff
@@ -0,0 +1,494 @@
+        (nonserialized.x0 == 0 && nonserialized.y0 == 0 &&
+        nonserialized.width + nonserialized.x0 >= nonserialized.img_width &&
+        nonserialized.height + nonserialized.y0 >= nonserialized.img_height)))]
+    source: u32,
+}
+
+struct RestorationFilterNonserialized {
+    encoding: Encoding,
+}
+
+#[derive(UnconditionalCoder, Debug, PartialEq)]
+#[nonserialized(RestorationFilterNonserialized)]
+struct RestorationFilter {
+    // all_default isn't mentioned in the spec, but libjxl has it
+    #[all_default]
+    #[default(true)]
+    all_default: bool,
+
+    #[default(true)]
+    gab: bool,
+
+    #[default(false)]
+    #[condition(gab)]
+    gab_custom: bool,
+
+    #[default(0.115169525)]
+    #[condition(gab_custom)]
+    gab_x_weight1: f32,
+
+    #[default(0.061248592)]
+    #[condition(gab_custom)]
+    gab_x_weight2: f32,
+
+    #[default(0.115169525)]
+    #[condition(gab_custom)]
+    gab_y_weight1: f32,
+
+    #[default(0.061248592)]
+    #[condition(gab_custom)]
+    gab_y_weight2: f32,
+
+    #[default(0.115169525)]
+    #[condition(gab_custom)]
+    gab_b_weight1: f32,
+
+    #[default(0.061248592)]
+    #[condition(gab_custom)]
+    gab_b_weight2: f32,
+
+    #[coder(Bits(2))]
+    #[default(2)]
+    epf_iters: u32,
+
+    #[default(false)]
+    #[condition(epf_iters > 0 && nonserialized.encoding == Encoding::VarDCT)]
+    epf_sharp_custom: bool,
+
+    #[default([0.0, 1.0 / 7.0, 2.0 / 7.0, 3.0 / 7.0, 4.0 / 7.0, 5.0 / 7.0, 6.0 / 7.0, 1.0])]
+    #[condition(epf_sharp_custom)]
+    epf_sharp_lut: [f32; 8],
+
+    #[default(false)]
+    #[condition(epf_iters > 0)]
+    epf_weight_custom: bool,
+
+    #[default([40.0, 5.0, 3.5])]
+    #[condition(epf_weight_custom)]
+    epf_channel_scale: [f32; 3],
+
+    #[default(0.45)]
+    #[condition(epf_weight_custom)]
+    epf_pass1_zeroflush: f32,
+
+    #[default(0.6)]
+    #[condition(epf_weight_custom)]
+    epf_pass2_zeroflush: f32,
+
+    #[default(false)]
+    #[condition(epf_iters > 0)]
+    epf_sigma_custom: bool,
+
+    #[default(0.46)]
+    #[condition(epf_sigma_custom && nonserialized.encoding == Encoding::VarDCT)]
+    epf_quant_mul: f32,
+
+    #[default(0.9)]
+    #[condition(epf_sigma_custom)]
+    epf_pass0_sigma_scale: f32,
+
+    #[default(6.5)]
+    #[condition(epf_sigma_custom)]
+    epf_pass2_sigma_scale: f32,
+
+    #[default(2.0 / 3.0)]
+    #[condition(epf_sigma_custom)]
+    epf_border_sad_mul: f32,
+
+    #[default(1.0)]
+    #[condition(epf_iters > 0 && nonserialized.encoding == Encoding::Modular)]
+    epf_sigma_for_modular: f32,
+
+    #[default(Extensions::default())]
+    extensions: Extensions,
+}
+
+pub struct FrameHeaderNonserialized {
+    pub xyb_encoded: bool,
+    pub num_extra_channels: u32,
+    pub have_animation: bool,
+    pub have_timecode: bool,
+    pub img_width: u32,
+    pub img_height: u32,
+}
+
+#[derive(UnconditionalCoder, Debug, PartialEq)]
+#[nonserialized(FrameHeaderNonserialized)]
+#[aligned]
+pub struct FrameHeader {
```

Comment:
I believe `#[validate]` should also be here (possibly as a follow up) - at least for validating extra channel upsampling.